### PR TITLE
Changes for compiling on Windows with MSys2.

### PR DESCRIPTION
**Pull request by hrittich**
- Included checks for the headers pwd.h, sys/utsname.h, direct.h.
- Included check for the sigaction function.
- Created utility mkdir function.
- Added -no-undefined flag to libtool.

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -231,6 +231,12 @@ LIBMESH_TEST_CXX11_NOEXCEPT
 #-----------------------------------------------------
 LT_INIT([disable-static])
 
+# We need to set the no-undefined flag for linking 
+# with libtool, because Windows requires all shared
+# libraries to include all used symbols. This flag
+# should do no harm on other platforms.
+libmesh_LDFLAGS+="-no-undefined"
```
Comment (pbauman) [inline configure.ac]: This broke compiling on Linux with g++ (the GRINS configure failure).

Comment (jwpeterson) [inline configure.ac]: Oddly, it passed on the INL Linux/GCC builds (Ubuntu 14.04 with hand-built GCC 6.2.0).

Comment (pbauman) [inline configure.ac]: I should've noted, libMesh built fine, but GRINS couldn't link against libMesh with that flag in the LD_FLAGS (which we snag from libmesh-config). Agreed it's odd.

**Comment by jwpeterson** [comment]
Our system wasn't quite happy with the `mkdir` changes though:
```
Function mkdir not available on this system.
Stack frames: 7
0: libMesh::print_trace(std::ostream&)
1: libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*)
2: libMesh::Utility::mkdir(char const*)
3: libMesh::RBEvaluation::legacy_write_offline_data_to_files(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)
4: /home/moosetest/civet/client_root_0/libmesh/build/examples/reduced_basis/reduced_basis_ex1/.libs/lt-example-opt() [0x414074]
5: __libc_start_main
6: /home/moosetest/civet/client_root_0/libmesh/build/examples/reduced_basis/reduced_basis_ex1/.libs/lt-example-opt() [0x414db4]
[0] ../src/utils/utility.C, line 146, compiled Aug  2 2017 at 08:26:30
```

File: configure.ac (hunk 1 above)
Comment (jwpeterson) [inline configure.ac]: Oh I see. The INL testing didn't make it past `make check` in libmesh, so we may have eventually had the same problem...

Comment (pbauman) [inline configure.ac]: Here's the full error message:

<pre>
configure:21001: checking for libMesh linkage
configure:21021: mpic++ -o conftest -O0 -felide-constructors -g -pedantic -W -Wall -Wextra -Wno-long-long -Wunused -Wpointer-arith -Wformat -Wparentheses -Woverloaded-virtual -fopenmp -g -O2 -I/femputer/pbauman/civet_build_testing_root/libmesh-dbg-install/include -I/femputer/pbauman/swstack/libs/hdf5/1.8.16/gcc/6.2.0/include -I/femputer/pbauman/swstack/libs/vtk-dbg/7.1.0/gcc/6.2.0/mpich/3.1.4/include/vtk-7.1 -pthread -I/femputer/pbauman/swstack/libs/petsc/3.7.4/gcc/6.2.0/mpich/3.1.4/openblas/0.2.18/opt/include -I/femputer/pbauman/swstack/libs/petsc/3.7.4/gcc/6.2.0/mpich/3.1.4/openblas/0.2.18/opt//include -I/femputer/pbauman/swstack/libs/mpich/3.1.4/gcc/6.2.0/include -I/femputer/pbauman/swstack/libs/boost-dbg/1.61.0/gcc/6.2.0/include -DDEBUG -D_GLIBCXX_DEBUG -D_GLIBCXX_DEBUG_PEDANTIC  -no-undefined  conftest.cpp -L/femputer/pbauman/swstack/libs/hdf5/1.8.16/gcc/6.2.0/lib -lhdf5 -lhdf5_cpp -Wl,-rpath,/femputer/pbauman/swstack/libs/hdf5/1.8.16/gcc/6.2.0/lib -Wl,-rpath,/femputer/pbauman/swstack/libs/vtk-dbg/7.1.0/gcc/6.2.0/mpich/3.1.4/lib -L/femputer/pbauman/swstack/libs/vtk-dbg/7.1.0/gcc/6.2.0/mpich/3.1.4/lib -lvtkIOCore-7.1 -lvtkCommonCore-7.1 -lvtkCommonDataModel-7.1 -lvtkFiltersCore-7.1 -lvtkIOXML-7.1 -lvtkImagingCore-7.1 -lvtkIOImage-7.1 -lvtkImagingMath-7.1 -lvtkIOParallelXML-7.1 -lvtkParallelMPI-7.1 -lvtkParallelCore-7.1 -lz -L/femputer/pbauman/swstack/libs/petsc/3.7.4/gcc/6.2.0/mpich/3.1.4/openblas/0.2.18/opt/lib -lpetsc -Wl,-rpath,/femputer/pbauman/swstack/libs/petsc/3.7.4/gcc/6.2.0/mpich/3.1.4/openblas/0.2.18/opt/lib -lsuperlu_dist -lcmumps -ldmumps -lsmumps -lzmumps -lmumps_common -lpord -lparmetis -lmetis -lscalapack -Wl,-rpath,/femputer/pbauman/swstack/libs/petsc/3.7.4/gcc/6.2.0/mpich/3.1.4/openblas/0.2.18/opt/lib64 -L/femputer/pbauman/swstack/libs/petsc/3.7.4/gcc/6.2.0/mpich/3.1.4/openblas/0.2.18/opt/lib64 -lsuperlu -lHYPRE -L/femputer/pbauman/swstack/libs/mpich/3.1.4/gcc/6.2.0/lib -L/femputer/pbauman/swstack/apps/gcc/6.2.0/lib/gcc/x86_64-pc-linux-gnu/6.2.0 -L/femputer/pbauman/swstack/apps/gcc/6.2.0/lib64 -L/femputer/pbauman/swstack/apps/gcc/6.2.0/lib -Wl,-rpath,/femputer/pbauman/swstack/libs/openblas/0.2.18/gcc/6.2.0/lib -L/femputer/pbauman/swstack/libs/openblas/0.2.18/gcc/6.2.0/lib -lopenblas -lmpifort -lgfortran -lquadmath -lm -lmpicxx -lstdc++ -Wl,-rpath,/femputer/pbauman/swstack/apps/gcc/6.2.0/lib/gcc/x86_64-pc-linux-gnu/6.2.0 -Wl,-rpath,/femputer/pbauman/swstack/apps/gcc/6.2.0/lib64 -Wl,-rpath,/femputer/pbauman/swstack/apps/gcc/6.2.0/lib -Wl,-rpath,/femputer/pbauman/swstack/libs/mpich/3.1.4/gcc/6.2.0/lib -lmpi -lgcc_s -ldl -Wl,-rpath,/femputer/pbauman/civet_build_testing_root/libmesh-dbg-install/lib -L/femputer/pbauman/civet_build_testing_root/libmesh-dbg-install/lib -lmesh_dbg  >&5
g++: error: unrecognized command line option '-no-undefined'; did you mean '-Wno-undef'?
</pre>

Hand-built GCC 6.2.0 on CentOS 6.

Comment (hrittich) [inline configure.ac]: The flag -no-undefined is a libtool flag. It looks like it is passed directly to the liker without libtool.

**Comment by jwpeterson** [comment]
Sorry, the CI was "waiting for a developer to activate". I sent you an invite to join the Associates team so your PR will automatically be tested @hrittich.

**Comment by hrittich** [comment]
@jwpeterson Thank you.

**Comment by jwpeterson** [comment]
Overall this looks great, I just have a few minor nitpicks that I will fix before merge (patch adds trailing whitespace to some files, need to split the bootstrapped files into a separate commit). It will be nice to be able to point to this PR when people ask about Windows support in the future!

**Comment by roystgnr** [comment]
:+1:, and many thanks!

**Comment by jwpeterson** [comment]
Merged in d940c39.

**Comment by hrittich** [comment]
Sounds great.

Concerning the PR. Currently, not all optional dependencies build under Windows. For example _fparser_ and _netcdf_ do not compile. The core library, however, compiles (using [msys2](http://www.msys2.org/)) and works. The following is the configuration that I have tested. 

    ./configure \
	--disable-trilinos \
	--enable-nodeconstraint  \
	--disable-mpi \
	--disable-amr \
	--disable-boost \
	--disable-petsc --disable-slepc \
	--disable-metis \
	--disable-parmetis \
	--disable-tetgen \
	--disable-hdf5 \
	--disable-netcdf \
	--disable-netcdf-4 --disable-dap \
	--disable-qhull \
	--disable-fparser

There is currently still one problem. Symlinks do not work under Windows. Hence when cloning the git repository, the symlinks have to by replaced by copies of the directories they point to. (See [here](https://stackoverflow.com/questions/5917249/git-symlinks-in-windows) for more details). Hence, for a more Windows friendly version, the symlinks could be removed from the repository.

**Comment by jwpeterson** [comment]
It seems like `--disable-amr` shouldn't be necessary for Windows builds... do you recall what didn't work for that configuration?  Also, I  suspect there must be some version of MPI (and hence PETSc) that would work under Windows... having those two would give you quite a bit more functionality.

> Symlinks do not work under Windows.

Yikes.

>  Hence when cloning the git repository, the symlinks have to by replaced by copies of the directories they point to.

Did you try the "mklink" option suggested in that StackOverflow article? That would probably be preferable to copies since many of our links:
```
README
contrib/eigen/eigen
contrib/exodusii/v5.09
contrib/exodusii/v5.22
contrib/nemesis/v3.09
contrib/nemesis/v5.22
contrib/netcdf/v3
contrib/netcdf/v4
contrib/qhull/qhull
contrib/tecplot/binary/lib/x86_64-apple-darwin10.8.0
contrib/tecplot/binary/lib/x86_64-apple-darwin11.4.2
contrib/tecplot/binary/lib/x86_64-apple-darwin12.2.0
contrib/tecplot/binary/lib/x86_64-apple-darwin12.2.1
contrib/tecplot/binary/lib/x86_64-apple-darwin12.5.0
contrib/tecplot/binary/lib/x86_64-apple-darwin13.1.0
contrib/tecplot/tecio/include/TECIO.h
contrib/tecplot/tecio/include/TECXXX.h
examples/systems_of_equations/systems_of_equations_ex8/systems_of_equations_ex8.exo
```
are to entire source trees. I would not be opposed to adding a `contrib/bin` script that would set up such links on Windows systems if you feel like contributing it.

**Comment by hrittich** [comment]
I just deactivated AMR because I did not need it. It might just work. I just have not tested it.

I try to build such a script tomorrow. Maybe it does make sense to put a _compiling on Windows_ section somewhere in the documentation.

**Comment by dschwen** [comment]
For future reference: compiling with FParser should work now. I'm also attaching a patch which should make netcdf build on msys2

**Pull request closed without merging**